Composer/packages/extensions/obiformeditor/src/Form/widgets/LuEditorWidget.tsx

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import React from 'react';
+import { LuEditor } from '@bfc/code-editor';
+import debounce from 'lodash/debounce';
+import { LuIntentSection } from '@bfc/shared';
+import { LuFile, filterSectionDiagnostics } from '@bfc/indexers';
+
+import { FormContext } from '../types';
+
+interface LuEditorWidgetProps {
+  formContext: FormContext;
+  name: string;
+  height?: number | string;
+  onChange: (template?: string) => void;
+}
+
+export class LuEditorWidget extends React.Component<LuEditorWidgetProps> {
```

Comment:
Let's convert this to a function component and use hooks instead of class methods. 

I believe the `getDerivedStateFromProps` implementation is causing some unexpected behavior in editing the form. For example, data is getting overwritten at times, and I'm unable to add newlines to the bottom of an intent definition like I can in the inline LG editor. The behavior should be almost identical to the inline LG editor. See `LgEditorWidget.tsx` for its implementation. It should be used as a strong reference.

Reply:
As @yeze322 said, 
>This component has several problems. The usage of useEffect hook is anti-pattern meanwhile the data flow is not well designed (as an example, the onChange handler submits data to both local states and external API).

>The props and local states are mixed together, makes your component hard to maintain.

>You really need to rethink what's your scenario is and carefully define the shape of your input props to make sure you can fully leverage change detection mechanism.
> Though state management of Shell has historical problems, you can try the Class Component with lifecycle methods before Leilei's fix on Shell states.
> (I was also trapped in this problem before)

There is a bug about luEditor value in `luEditorWidget`: state `localVale` and props `luIntent.body` are mixed. So I try to use class component to fix it. 
I fixed the bug: unable to add newlines to the bottom.

Reply:
Understood. I cut this: https://github.com/microsoft/BotFramework-Composer/issues/2138 to follow-up.

Reply:
Under current data flow, `getDerivedStateFromProps` can fully support weitian's scenario though it's still not a good pattern.

I think the best practice of this scenario is to use a fully uncontrolled component with a `key` (according to [official doc](https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops)). A key here might be the data version of current local state, requires shell or indexer work